activation/nipost.go

```diff
@@ -158,22 +158,27 @@ func (nb *NIPostBuilder) BuildNIPost(ctx context.Context, challenge *types.NIPos
 
 	// Phase 0: Submit challenge to PoET services.
 	if nb.state.PoetRequests == nil {
+		now := time.Now()
+		if poetRoundStart.Before(now) {
+			return nil, 0, fmt.Errorf("%w: poet round has already started at %s (now: %s)", ErrATXChallengeExpired, poetRoundStart, now)
+		}
+
 		signature := nb.signer.Sign(signing.POET, challengeHash.Bytes())
 		prefix := bytes.Join([][]byte{nb.signer.Prefix(), {byte(signing.POET)}}, nil)
 		submitCtx, cancel := context.WithDeadline(ctx, poetRoundStart)
 		defer cancel()
 		poetRequests := nb.submitPoetChallenges(submitCtx, prefix, challengeHash.Bytes(), signature, nb.signer.NodeID())
-		if err := ctx.Err(); err != nil {
-			return nil, 0, fmt.Errorf("submitting challenges: %w", err)
-		}
-
 		if len(poetRequests) == 0 {
-			return nil, 0, &PoetSvcUnstableError{msg: "failed to submit challenge to any PoET"}
+			return nil, 0, &PoetSvcUnstableError{msg: "failed to submit challenge to any PoET", source: ctx.Err()}
 		}
+
 		nipost.Challenge = &challengeHash
 		nb.state.Challenge = challengeHash
 		nb.state.PoetRequests = poetRequests
 		nb.persist()
```

Comment:
does this answer @dshulyak's question [here](https://github.com/spacemeshos/go-spacemesh/pull/4259#issuecomment-1497766960) ?

Reply:
thanks, i wasn't sure that they are persisted with a challenge